### ISSUE TITLE
Strange F821 false positive on flake8 git main

### ISSUE THREAD

**Issue by nyuszika7h** [snok/flake8-type-checking#71]
When using this extension with PyCQA/flake8@8c1ed24, it triggers false F821 errors from pyflakes. The error does not happen with flake8 4.0.1 from PyPI. It only seems to happen on top-level code, ~~and goes away if a non-empty `TYPE_CHECKING` block with any import (even an unused one) is present~~.

# Minimal working example

```py
import os

os.urandom(8)
```

pyproject.toml:
```toml
[tool.poetry]
name = "test"
version = "0.1.0"
description = ""
authors = ["nyuszika7h <nyuszika7h@gmail.com>"]

[tool.poetry.dependencies]
python = "^3.10"

[tool.poetry.dev-dependencies]
flake8 = {git = "https://github.com/PyCQA/flake8.git", rev = "main"}
flake8-type-checking = "^1.3.1"

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

flake8 output:
```
❯ flake8 test.py
test.py:3:1: F821 undefined name 'os'
```

---

~~The following does not trigger the error:~~
```py
import os                                               
from typing import TYPE_CHECKING                        
                                                        
if TYPE_CHECKING:                                       
    import sys  # random import to demonstrate, not used
                                                        
os.urandom(8)                                           
```

**Comment by sondrelg** [comment]
Is the error not present when you remove `flake8-type-checking`?

**Comment by nyuszika7h** [comment]
Yes, it doesn't appear if I remove it:

```
❯ flake8 test.py
test.py:4:4: F821 undefined name 'TYPE_CHECKING'
test.py:7:1: F821 undefined name 'os'
❯ pip uninstall flake8-type-checking
Found existing installation: flake8-type-checking 1.3.1
Uninstalling flake8-type-checking-1.3.1:
  Would remove:
    /home/nyuszika7h/pytest/.venv/lib/python3.10/site-packages/flake8_type_checking-1.3.1.dist-info/*
    /home/nyuszika7h/pytest/.venv/lib/python3.10/site-packages/flake8_type_checking/*
Proceed (Y/n)? y
  Successfully uninstalled flake8-type-checking-1.3.1
❯ flake8 test.py
test.py:5:5: F401 'sys' imported but unused
```

(Strangely, now it's even complaining about `TYPE_CHECKING` when I didn't before... Also, you may have to apply #70 first to even see these errors.)

**Comment by sondrelg** [comment]
Ok good to know. I'll try to take a look at this this week. If you want to submit a PR in the meantime, that's of course always welcome 👍  Thanks for reporting it

**Comment by nyuszika7h** [comment]
Just to clarify, I found out the second example with the `TYPE_CHECKING` block was not triggering the error because of the error from #70 (which wasn't showing when I was just checking the automatic linter output from neovim), but after that was merged the F821 false positives happen on both examples.

**Comment by sondrelg** [comment]
Would you mind checking if it still fails for you with v1.3.3 @nyuszika7h? I think it works for me now, but I might have tested it wrong.

**Comment by sondrelg** [comment]
I'm going to close this for now, assuming it is fixed 👍 Let me know if that's not right

**Comment by nyuszika7h** [comment]
Sorry for the late reply. I can still reproduce the issue:

```
❯ mkcd flake8-type-checking-issue-71
❯ poetry --version
Poetry version 1.1.13
❯ poetry init -n

This command will guide you through creating your pyproject.toml config.


You can specify a package in the following forms:
  - A single name (requests)
  - A name and a constraint (requests@^2.23.0)
  - A git url (git+https://github.com/python-poetry/poetry.git)
  - A git url with a revision (git+https://github.com/python-poetry/poetry.git#develop)
  - A file path (../my-package/my-package.whl)
  - A directory (../my-package/)
  - A url (https://example.com/packages/my-package-0.1.0.tar.gz)



❯ poetry add --dev git+https://github.com/PyCQA/flake8.git@50d6915 flake8-type-checking
Creating virtualenv flake8-type-checking-issue-71 in /home/nyuszika7h/flake8-type-checking-issue-71/.venv
Using version ^1.3.3 for flake8-type-checking

Updating dependencies
Resolving dependencies... (12.3s)

Writing lock file

Package operations: 5 installs, 0 updates, 0 removals

  • Installing mccabe (0.6.1)
  • Installing pycodestyle (2.8.0)
  • Installing pyflakes (2.4.0)
  • Installing flake8 (4.0.1 50d6915)
  • Installing flake8-type-checking (1.3.3)
❯ printf 'import os\n\nos.urandom(8)\n' > test.py
❯ flake8 test.py
test.py:3:1: F821 undefined name 'os'
```

Also, there's a new strange error when running flake8 on the whole project. Should I open a separate issue for this?

<details>
<summary>Error with <code>flake8 .</code></summary>

```
❯ flake8 .
/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/_distutils_hack/__init__.py:17: UserWarning: Distutils was imported before Setuptools, but importing Setuptools also replaces the `distutils` module in `sys.modules`. This may lead to undesirable behaviors or errors. To avoid these issues, avoid using distutils directly, ensure that setuptools is installed in the traditional way (e.g. not an editable install), and/or make sure that setuptools is always imported before distutils.
  warnings.warn(
/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/checker.py", line 623, in _run_checks
    return checker.run_checks()
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/checker.py", line 533, in run_checks
    self.run_ast_checks()
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/checker.py", line 436, in run_ast_checks
    for (line_number, offset, text, _) in runner:
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8_type_checking/plugin.py", line 81, in run
    visitor = TypingOnlyImportsChecker(self._tree, self.options)
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8_type_checking/checker.py", line 568, in __init__
    self.visitor.visit(node)
  File "/home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/ast.py", line 410, in visit
    return visitor(node)
  File "/home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/ast.py", line 486, in generic_visit
    value = self.visit(value)
  File "/home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/ast.py", line 410, in visit
    return visitor(node)
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8_type_checking/checker.py", line 321, in visit_ImportFrom
    self._add_import(node)
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8_type_checking/checker.py", line 302, in _add_import
    is_local = self._import_is_local(f'{module}{name_node.name}')
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8_type_checking/checker.py", line 227, in _import_is_local
    spec = find_spec('.'.join(import_name.split('.')[:-1]), import_name.split('.')[-1])
  File "/home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/importlib/util.py", line 94, in find_spec
    parent = __import__(parent_name, fromlist=['__path__'])
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/setuptools/__init__.py", line 8, in <module>
    import _distutils_hack.override  # noqa: F401
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/_distutils_hack/override.py", line 1, in <module>
    __import__('_distutils_hack').do_override()
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 72, in do_override
    ensure_local_distutils()
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 59, in ensure_local_distutils
    assert '_distutils' in core.__file__, core.__file__
AssertionError: /home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/distutils/core.py
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/bin/flake8", line 8, in <module>
    sys.exit(main())
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/main/cli.py", line 22, in main
    app.run(argv)
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/main/application.py", line 334, in run
    self._run(argv)
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/main/application.py", line 323, in _run
    self.run_checks()
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/main/application.py", line 232, in run_checks
    self.file_checker_manager.run()
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/checker.py", line 253, in run
    self.run_parallel()
  File "/home/nyuszika7h/flake8-type-checking-issue-71/.venv/lib/python3.10/site-packages/flake8/checker.py", line 219, in run_parallel
    for ret in pool_map:
  File "/home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/multiprocessing/pool.py", line 448, in <genexpr>
    return (item for chunk in result for item in chunk)
  File "/home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/multiprocessing/pool.py", line 870, in next
    raise value
AssertionError: /home/nyuszika7h/.pyenv/versions/3.10.2/lib/python3.10/distutils/core.py
```

</details>

**Comment by sondrelg** [comment]
Thanks for checking @nyuszika7h. You don't need to open a separate issue, I'll try to replicate and fix both when I get a chance 👍

**Comment by sondrelg** [comment]
Think this should be resolved in v1.5.0. Seems like copying the ast-tree so modifications aren't propagated to other plugins solves it: https://github.com/snok/flake8-type-checking/pull/81
